goinstallscript/main.go

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"os"
+	"os/exec"
+
+	"github.com/microsoft/go-infra/goinstallscript/powershell"
+)
+
+// scriptEmitPackage is a Go package that emits the PowerShell script content. The goinstallscript
```

Comment:
I'm afraid I don't get the use-case that this indirection tries to tackle. Why someone would want to install multiple versions of the script? How other steps would be necessary to accomplish that?

Also, this new `go run` approach precludes running `goinstallscript` in environments without `go` installed. I think it was a nice feature that before you could plant `goinstallscript` in a brand new environament and just execute it to install the powershell script. Note that we could support both use-cases by falling back to `powershell.Content` if [exec.LookPath("go")](https://pkg.go.dev/os/exec#LookPath).

Reply:
The main scenario here is:

1. Dev installs github.com/microsoft/go-infra/goinstallscript v0.0.1.
2. Dev sets up their repo with v0.0.1 of the script.
3. Some time passes.
4. We update the script, dependabot submits a PR to v0.0.2 of github.com/microsoft/go-infra/goinstallscript.
5. CI fails because the v0.0.1 content is still in the repo.
6. Dev gets the branch and runs `goinstallscript`...

* Without indirection, `goinstallscript` is still `v0.0.1`, so it puts the `v0.0.1` content in the file again, accomplishing nothing.
* With indirection, they get the `v0.0.2` content without having to reinstall `goinstallscript`. They're able to push the change back up to the repo and move on.

Some other similar mixups could happen if the dev works on multiple repos, e.g. one uses `v0.0.2` of the script and another uses `v0.0.3`--relying on the user-wide `goinstallscript`'s embedded content means they need to run `go install` very often.

This is all about https://github.com/microsoft/go-infra/pull/182#discussion_r1806762461.

But... now that I'm faced with this again, I'm realizing that we can go back to the direct package and simply document that people should use `go run`, not `go install`. I didn't go back to reevaluate once I realized `go run` is useful here. 😄 

---

FWIW, so far, I've only considered the scenario of using the *script* to install Go without Go, and `goinstallscript` is purely for someone who's trying to set up CI in a Go repo they've been working on (so they would already have Go). I'm not able to think of a scenario where it's reasonable to transfer a prebuilt `goinstallscript` binary to a machine and not ok to transfer `go-install.ps1` itself, though...

Reply:
@qmuntal, does the use case and simplification make sense?